### PR TITLE
`fn dav1d_parse_sequence_header`: backport simplification from dav1d 1.2.1

### PR DESCRIPTION
**Pull request by fbossen**
Avoid creating an entire decoder instance to just parse a sequence header.

### CODE REVIEW

File: src/obu.rs [hunk 1]
```diff
@@ -549,6 +541,47 @@ fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSeq
...
 
+pub(crate) unsafe fn rav1d_parse_sequence_header(
+    mut data: &[u8],
+) -> Rav1dResult<DRav1d<Rav1dSequenceHeader, Dav1dSequenceHeader>> {
+    while !data.is_empty() {
+        let mut gb = GetBits::new(data);
+
+        gb.get_bit(); // obu_forbidden_bit
+        let raw_type = gb.get_bits(4);
+        let r#type = Rav1dObuType::from_repr(raw_type as usize);
```
Comment (kkysen) [inline src/obu.rs:552]:
```suggestion
        let r#type = Rav1dObuType::from_repr(gb.get_bits(4) as usize).unwrap();
```

File: src/obu.rs [hunk 2]
```diff
@@ -549,6 +541,47 @@ fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSeq
...
+            let len = gb.get_uleb128() as usize;
+            if len > data.len() {
+                return Err(EINVAL);
+            }
+            len
+        } else {
+            data.len()
+        };
+
+        if r#type == Some(Rav1dObuType::SeqHdr) {
```
Comment (kkysen) [inline src/obu.rs:568]:
```suggestion
        if r#type == Rav1dObuType::SeqHdr {
```

File: src/obu.rs [hunk 3]
```diff
@@ -549,6 +541,47 @@ fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSeq
     })
 }
 
+pub(crate) unsafe fn rav1d_parse_sequence_header(
+    mut data: &[u8],
+) -> Rav1dResult<DRav1d<Rav1dSequenceHeader, Dav1dSequenceHeader>> {
+    while !data.is_empty() {
```
Comment (kkysen) [inline src/obu.rs:547]:
```suggestion
    let res = Err(ENOENT);

    while !data.is_empty() {
```

File: src/obu.rs [hunk 4]
```diff
@@ -549,6 +541,47 @@ fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSeq
...
+        }
+
+        if gb.has_error() != 0 {
+            return Err(EINVAL);
+        }
+
+        data = &data[obu_end..]
+    }
+
+    Err(ENOENT)
```
Comment (kkysen) [inline src/obu.rs:582]:
```suggestion
    res.map(DRav1d::from_rav1d)
```

File: src/obu.rs [hunk 5]
```diff
@@ -549,6 +541,47 @@ fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSeq
...
+            }
+            len
+        } else {
+            data.len()
+        };
+
+        if r#type == Some(Rav1dObuType::SeqHdr) {
+            if let Ok(seq_hdr) = parse_seq_hdr(&mut gb, false) {
+                return Ok(DRav1d::from_rav1d(seq_hdr));
+            }
```
Comment (kkysen) [inline src/obu.rs:571]:
```suggestion
            let res = parse_seq_hdr(&mut gb, false);
            res?;
```

File: src/obu.c [hunk 6]
```diff
@@ -283,6 +278,45 @@ static int parse_seq_hdr(Dav1dContext *const c, GetBits *const gb,
...
+        }
+
+        if (type == DAV1D_OBU_SEQ_HDR) {
+            if ((res = parse_seq_hdr(out, &gb, 0)) < 0) return res;
+            if (gb.ptr > obu_end) return DAV1D_ERR(EINVAL);
+            dav1d_bytealign_get_bits(&gb);
+        }
+
+        if (gb.error) return DAV1D_ERR(EINVAL);
+        assert(gb.state == 0 && gb.bits_left == 0);
```
Comment (kkysen) [inline src/obu.c:313]: I think you missed this line.

File: src/obu.c [hunk 7]
```diff
@@ -283,6 +278,45 @@ static int parse_seq_hdr(Dav1dContext *const c, GetBits *const gb,
...
+        const uint8_t *obu_end = gb.ptr_end;
+        if (has_length_field) {
+            const size_t len = dav1d_get_uleb128(&gb);
+            if (len > (size_t)(obu_end - gb.ptr)) return DAV1D_ERR(EINVAL);
+            obu_end = gb.ptr + len;
+        }
+
+        if (type == DAV1D_OBU_SEQ_HDR) {
+            if ((res = parse_seq_hdr(out, &gb, 0)) < 0) return res;
+            if (gb.ptr > obu_end) return DAV1D_ERR(EINVAL);
```
Comment (kkysen) [inline src/obu.c:308]: This line needs to be ported, too.

**Review by kkysen — CHANGES_REQUESTED** [review]

File: src/obu.rs (hunk 1 above)
Comment (fbossen) [reply src/obu.rs:552]: Not sure that's the right thing to do. Not all 16 possible values of `Rav1dObuType` are defined. As per the AV1 specification, values that aren't recognized should simply be ignored. We shouldn't panic when encountering one of the undefined values.

Comment (kkysen) [reply src/obu.rs]: Oh okay, I might've misread that.

File: src/obu.rs (hunk 3 above)
Comment (fbossen) [reply src/obu.rs:549]: I made `res` mutable so that it can be modified in the loop

File: src/obu.rs (hunk 5 above)
Comment (fbossen) [reply src/obu.rs]: I removed the `let` to avoid creating a shadow variable

File: src/obu.c (hunk 6 above)
Comment (fbossen) [reply src/obu.c:313]: Those are private fields, so I added a `has_pending_bits` method

File: src/obu.c (hunk 7 above)
Comment (fbossen) [reply src/obu.c:308]: Done

**Comment by fbossen** [comment]
Note sure whether the API function `dav1d_parse_sequence_header` is tested in CI. Ended up using `-s 5` when running the decoder to test.

File: src/obu.rs [hunk 8]
```diff
@@ -549,6 +541,51 @@ fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSeq
...
+            if gb.byte_pos() > obu_end {
+                return Err(EINVAL);
+            }
+            gb.bytealign();
+        }
+
+        if gb.has_error() != 0 {
+            return Err(EINVAL);
+        }
+        assert_eq!(gb.has_pending_bits(), false);
```
Comment (kkysen) [inline src/obu.rs:581]:
```suggestion
        assert!(!gb.has_pending_bits());
```

**Review by kkysen — APPROVED** [review]
LGTM!  Could you just use `assert!` instead of `assert_eq!` on a `bool`?

File: src/obu.rs (hunk 8 above)
Comment (fbossen) [reply src/obu.rs]: Done

File: src/getbits.rs [hunk 9]
```diff
@@ -189,4 +189,14 @@ impl<'a> GetBits<'a> {
     pub const fn pos(&self) -> usize {
         self.index * u8::BITS as usize - self.bits_left as usize
     }
+
+    #[inline]
```
Comment (kkysen) [inline src/getbits.rs:193]:
```suggestion
```

We shouldn't need manual `#[inline]`s where C doesn't have them and looking at the assembly doesn't show a difference.  If it's a short `const fn`, it's basically `#[inline]` anyways, and in practice, `rustc` now makes all short functions `#[inline]` (that's why we need `#[inline(never)]` on godbolt for short functions).

File: src/getbits.rs [hunk 10]
```diff
@@ -189,4 +189,14 @@ impl<'a> GetBits<'a> {
...
+    pub const fn byte_pos(&self) -> usize {
+        self.index
+    }
+
+    #[inline]
```
Comment (kkysen) [inline src/getbits.rs:198]:
```suggestion
```

File: src/obu.rs [hunk 11]
```diff
@@ -124,7 +126,10 @@ impl Debug {
     }
 }
 
-fn parse_seq_hdr(c: &mut Rav1dContext, gb: &mut GetBits) -> Rav1dResult<Rav1dSequenceHeader> {
+fn parse_seq_hdr(
```
Comment (kkysen) [inline src/obu.rs:129]:
```suggestion
#[inline(never)]
fn parse_seq_hdr(
```

Porting C's `NOINLINE` addition.

**Review by kkysen — CHANGES_REQUESTED** [review]
Sorry I missed these couple things before.  We should port the `NOINLINE` C added.

**Review by kkysen — APPROVED** [review]

**Pull request merged**